src/test/java/htsjdk/samtools/filter/TagFilterTest.java

```diff
@@ -38,41 +39,163 @@
-     * @param validValues       The values the filter should test for
-     * @param testValue         The value to test for in the record
-     * @param expectedResult    The expected result (true is the sequence should match the filter, otherwise false)
+     * @param commonValuesIndex         The index of the common values to grab from commonTestValues
+     * @param firstReadExpectedResult   The expected result of the first read (true is the sequence should match the filter, otherwise false)
+     * @param pairedExpectedResult      The expected result of the paired reads (true is the sequence should match the filter, otherwise false)
+     * @param includeReads              The value of includeReads for the filter
+     * @param matchPairs                Whether or not to have matching values for tag the test is testing for
+     *
      */
-    @Test(dataProvider="data")
-    public void testTagFilter(final String testName, final String tag, final List<Object> validValues,
-                              final Object testValue, final boolean expectedResult) {
-        final TagFilter filter = new TagFilter(tag, validValues);
-        builder.addUnmappedFragment("testfrag");
-        final SAMRecord record = builder.iterator().next();
+
+    private void testTagFilter(final int commonValuesIndex, final boolean firstReadExpectedResult,
+                               final boolean pairedExpectedResult, final Boolean includeReads,
+                               final boolean matchPairs) {
+
+        final String testName = (String) commonTestValues[commonValuesIndex][0];
+        final String tag = (String) commonTestValues[commonValuesIndex][1];
+        final List<Object> validValues = (List<Object>) commonTestValues[commonValuesIndex][2];
+        final Object testValue = commonTestValues[commonValuesIndex][3];
+
+        final TagFilter filter = new TagFilter(tag, validValues, includeReads);
+        final SAMRecordSetBuilder builder = new SAMRecordSetBuilder();
+        builder.addPair("Paired", 1, 100, 200);
+        CloseableIterator<SAMRecord> iterator = builder.iterator();
+        final SAMRecord record1 = iterator.next();
         if (testValue != null) {
-            record.setAttribute(tag, testValue);
+            record1.setAttribute(tag, testValue);
         }
-        Assert.assertEquals(filter.filterOut(record), expectedResult, testName);
+        // Test first read in pair
+        Assert.assertEquals(filter.filterOut(record1), firstReadExpectedResult, testName);
+
+        final SAMRecord record2 = iterator.next();
+        if (matchPairs && testValue != null) {
+            record2.setAttribute(tag, testValue);
+        } else if (!matchPairs){
+            record2.setAttribute(tag, 0);
+        }
+        // Test paired reads
+        Assert.assertEquals(filter.filterOut(record1, record2), pairedExpectedResult, testName);
+    }
+
+    @Test(dataProvider="dataDefaultMatchingPairedFilter")

```

Comment:
These are all the same test! why not join all the data-providers and put comments describing what each "block" does?

Reply:
if i combine them its gonna be super annoying to figure out which test failed if one does

Reply:
hmmm. we have large data-providers all the time...it's not a problem...but if you like, you can include a String "name" in the input data and include it in the assertion. repeated tests that only differ by their names are worse (IMHO)

Reply:
I mean a test that "only differ by their names" mean the values woudl be the same...in which this case they are not.  Could you point me to one of these large data providers.

Reply:
i just don't think having to sift through 24 tests to figure out where the ` true != false` test failed seems like the most fun thing ever but if its done everywhere then i can change it.

Reply:
take a look at the tests for LiftOverVcf in picard for example.

Reply:
no sifting required...in the data object, add a string that decribes the data "DefaultMatchingPairedFilter" or whatever. and if the test fails include that string in the exception....

Reply:
ok